src/box/relay.cc

```diff
@@ -353,9 +353,8 @@ relay_cancel(struct relay *relay)
 {
 	/* Check that the thread is running first. */
 	if (relay->cord.id != 0) {
-		if (tt_pthread_cancel(relay->cord.id) == ESRCH)
-			return;
-		tt_pthread_join(relay->cord.id, NULL);
+		cord_cancel_and_join(&relay->cord);
+		relay->cord.id = 0;
```

Comment:
> you wanted to emphasize that this is the place where cord.id becomes 0

Yes.

> I suggest to wrap cord.id == 0 into a helper then

Nah, the problem right now is that `cord_join/cojoin/cancel_and_join()` all destroy the cord. So they not only stop it but also literally free it, turn into garbage. First two really call `cord_destroy()`. Even if `tt_pthread_join()` fails btw.

So calling any methods on the cord after join isn't really safe, like assume the id should become zero or something else.

While looking into that I discovered another one - we have no constructor which could create the cord but not start it yet. If we want this `cord_is_running()`, it would require to fix all those issues first, and I didn't want to do this in scope of this patch. Also I was too lazy to create a ticket, because it wouldn't even be a bugfix but just refactoring. PMs would put it to wishlist forever. I was hoping it might be done alongside fixing some other bug.

Reply:
Ok, I see your point, thanks.